src/views/Generator/RuleEditor/Typeahead/useTypeahead.tsx

```diff
@@ -0,0 +1,178 @@
+      : extDefaultValue != null
+        ? String(extDefaultValue)
+        : ''
+  )
+
+  React.useEffect(() => {
+    if (isControlled) {
+      setInputValue(String(extValue))
+    }
+  }, [extValue, isControlled])
+
+  const [isFocused, setFocused] = React.useState(false)
+  const [activeIndex, setActiveIndex] = React.useState<number | null>(null)
+
+  const shouldSuggest = React.useMemo(() => {
+    if (mode === 'onDot') return inputValue.includes('.')
+    if (mode === 'onFirstKey') return inputValue.length >= 1
+    if (mode === 'onThirdKey') return inputValue.length >= 3
+    return false
+  }, [inputValue, mode])
+
+  const filteredOptions = React.useMemo(() => {
+    if (!shouldSuggest) return []
+    if (mode === 'onDot') {
+      const parts = inputValue.split('.')
+      const prefix = parts.slice(0, -1).join('.')
+      const current = parts[parts.length - 1] || ''
+      const base = prefix ? `${prefix}.` : ''
+      return options.filter(
+        (o) =>
+          o.startsWith(base) &&
+          o.slice(base.length).toLowerCase().startsWith(current.toLowerCase())
+      )
+    }
+    return options.filter((o) =>
+      o.toLowerCase().includes(inputValue.toLowerCase())
+    )
+  }, [options, inputValue, shouldSuggest, mode])
+
+  const selectOption = React.useCallback(
+    (value: string) => {
+      if (!isControlled) {
+        setInputValue(value)
+      }
+      extOnChange?.({
+        target: {
+          name: restProps.name,
+          value,
+        },
+      } as React.ChangeEvent<HTMLInputElement>)
+      setFocused(false)
+    },
+    [extOnChange, isControlled, restProps.name]
+  )
+
+  const onKeyDown = React.useCallback(
+    (e: React.KeyboardEvent<HTMLInputElement>) => {
+      if (filteredOptions.length === 0) return
+      switch (e.key) {
+        case 'ArrowDown':
+          e.preventDefault()
+          setActiveIndex((i) =>
+            i === null || i === filteredOptions.length - 1 ? 0 : i + 1
+          )
+          break
+        case 'ArrowUp':
+          e.preventDefault()
+          setActiveIndex((i) =>
+            i === null || i === 0 ? filteredOptions.length - 1 : i - 1
+          )
+          break
+        case 'Enter':
+        case 'Tab':
+          if (activeIndex !== null) {
+            e.preventDefault()
+            if (filteredOptions[activeIndex]) {
+              selectOption(filteredOptions[activeIndex])
+            }
+          }
+          break
+        case 'Escape':
+          e.preventDefault()
+          setFocused(false)
+          break
+      }
+    },
+    [filteredOptions, activeIndex, selectOption]
+  )
+
+  const handleChange: React.ChangeEventHandler<HTMLInputElement> = (e) => {
+    const v = e.target.value ?? ''
+    if (!isControlled) {
+      setInputValue(v)
+    }
+    extOnChange?.(e)
+    const shouldFocus =
+      (mode === 'onDot' && v.includes('.')) ||
+      (mode === 'onFirstKey' && v.length >= 1) ||
+      (mode === 'onThirdKey' && v.length >= 3)
```

Comment:
It seems like this logic overlaps with `shouldSuggest`. What do you think about combining them?